src/rust/engine/process_execution/src/nailgun/mod.rs

```diff
@@ -0,0 +1,254 @@
+use std::collections::btree_map::BTreeMap;
+use std::collections::btree_set::BTreeSet;
+use std::os::unix::fs::symlink;
+use std::path::PathBuf;
+use std::sync::Arc;
+use std::time::Duration;
+
+use futures::future::Future;
+use log::{debug, trace};
+
+use boxfuture::{Boxable, BoxFuture, try_future};
+use hashing::Digest;
+use workunit_store::WorkUnitStore;
+
+use crate::{ExecuteProcessRequest, ExecuteProcessRequestMetadata, FallibleExecuteProcessResult, MultiPlatformExecuteProcessRequest, Platform};
+use crate::nailgun::nailgun_pool::NailgunProcessName;
+
+pub mod nailgun_pool;
+
+pub type NailgunPool = nailgun_pool::NailgunPool;
+
+// Hardcoded constants for connecting to nailgun
+static NAILGUN_MAIN_CLASS: &str = "com.martiansoftware.nailgun.NGServer";
+static ARGS_TO_START_NAILGUN: [&str; 1] = [":0"];
+
+// We can hardcode this because we mix it into the digest in the EPR.
+static NG_CLIENT_PATH: &str = "bin/ng/1.0.0/ng";
+
+/// Represents the result of parsing the args of a nailgunnable ExecuteProcessRequest
+/// TODO We may want to split the classpath by the ":", and store it as a Vec<String>
+///         to allow for deep fingerprinting.
+struct ParsedArgLists {
+    nailgun_args: Vec<String>,
+    client_args: Vec<String>,
+}
+
+///
+/// Given a list of args that one would likely pass to a java call,
+/// we automatically split it to generate two argument lists:
+///  - nailgun arguments: The list of arguments needed to start the nailgun server.
+///    These arguments include everything in the arg list up to (but not including) the main class.
+///    These arguments represent roughly JVM options (-Xmx...), and the classpath (-cp ...).
+///
+///  - client arguments: The list of arguments that will be used to run the jvm program under nailgun.
+///    These arguments can be thought of as "passthrough args" that are sent to the jvm via the nailgun client.
+///    These arguments include everything starting from the main class.
+///
+/// We assume that:
+///  - Every args list has a main class.
+///  - There is exactly one argument that doesn't begin with a `-` in the command line before the main class,
+///    and it's the value of the classpath (i.e. `-cp scala-library.jar`).
+///
+/// We think these assumptions are valid as per: https://github.com/pantsbuild/pants/issues/8387
+///
+fn split_args(args: &Vec<String>) -> ParsedArgLists {
```

Comment:
Optional: I understanding you are iterating right now, but having tests may be beneficial to minimize surprises during the iterations and makes the behavior more obvious.

Reply:
I have refactored that into its own module, turned it into a proper parser, and added tests: https://github.com/pantsbuild/pants/pull/8371/files#diff-41bbdca8880c484c302ec4b79dcd3aa8